akvo/rsr/usecases/jobs/aggregation.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import annotations
+from django.db.models import QuerySet
+from django.db.transaction import atomic
+from typing import List, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from akvo.rsr.models import IndicatorPeriod
+
+from akvo.rsr.models.aggregation_job import IndicatorPeriodAggregationJob
+from akvo.rsr.models.cron_job import CronJobMixin
+from akvo.rsr.usecases.period_update_aggregation import aggregate
+from akvo.rsr.usecases.jobs.cron import is_job_dead
+
+
+def get_scheduled_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return IndicatorPeriodAggregationJob.objects.filter(
+        status=CronJobMixin.Status.SCHEDULED,
+    )
+
+
+def get_running_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return IndicatorPeriodAggregationJob.objects.filter(
+        status=CronJobMixin.Status.RUNNING,
+    )
+
+
+def get_failed_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return IndicatorPeriodAggregationJob.objects.filter(
+        status=CronJobMixin.Status.FAILED,
+    )
+
+
+def get_maxxed_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return IndicatorPeriodAggregationJob.objects.filter(
+        status=CronJobMixin.Status.MAXXED,
+    )
+
+
+def get_finished_jobs() -> QuerySet[IndicatorPeriodAggregationJob]:
+    return IndicatorPeriodAggregationJob.objects.filter(
+        status=CronJobMixin.Status.FINISHED,
+    )
+
+
+def schedule_aggregation_job(period: IndicatorPeriod) -> IndicatorPeriodAggregationJob:
+    """
+    Schedule a job for the period to be aggregated upwards if no job exists
+    """
+    if existing_job := get_scheduled_jobs().filter(period=period).first():
+        existing_job.save()
+        return existing_job
+
+    program = period.indicator.result.project.ancestor()
+    return IndicatorPeriodAggregationJob.objects.create(period=period, program=program)
+
+
+def execute_aggregation_jobs():
+    """
+    Call the aggregation function for each aggregation job
+    """
+    handle_failed_jobs()
+
+    scheduled_jobs = get_scheduled_jobs()
+    for scheduled_job in scheduled_jobs:
+        with atomic():
+            scheduled_job.mark_running()
+            try:
+                run_aggregation(scheduled_job.period)
+                scheduled_job.mark_finished()
+            except Exception as e:
+                scheduled_job.mark_failed()
+                email_failed_job_owner(scheduled_job, str(e))
+
+
+@atomic
+def handle_failed_jobs():
+    """Identify failed jobs, notify owners, and reschedule them"""
+    fail_dead_jobs()
+    failed_jobs = get_failed_jobs()
```

Comment:
The failed jobs that are not caused by a dead process could also be included

Reply:
They are. `fail_dead_jobs()` marks the dead jobs as failed so they are included in the results of all other failed jobs to be rescheduled.